File: its-my-training-day/leetcode-top-75/questions/trees/tree-dfs/binary-tree-max-path-sum/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def binary_tree_max_path_sum(root: Optional[TreeNode]) -> int:
    best = -10**9

    def dfs(node: Optional[TreeNode]) -> int:
        nonlocal best
        if node is None:
            return 0
        left = dfs(node.left)
        right = dfs(node.right)
        left = left if left > 0 else 0
        right = right if right > 0 else 0
        total = node.val + left + right
        if total > best:
            best = total
        return node.val + (left if left > right else right)

    dfs(root)
    return best

File: its-my-training-day/leetcode-top-75/questions/trees/tree-dfs/binary-tree-max-path-sum/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const binaryTreeMaxPathSum = (root: TreeNode | null): number => {
  let best = -1e9;
  const dfs = (node: TreeNode | null): number => {
    if (node === null) return 0;
    let left = dfs(node.left);
    let right = dfs(node.right);
    if (left < 0) left = 0;
    if (right < 0) right = 0;
    const total = node.val + left + right;
    if (total > best) best = total;
    return node.val + (left > right ? left : right);
  };
  dfs(root);
  return best;
};
